File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.cpp
class Solution {
  public:
    int longestSubarray(vector<int>& arr, int k) {
        // code here
        map<int,int> mp;
        int maxi = 0;
        int sum = 0;
        for(int i=0;i<arr.size();i++){
            sum+=arr[i];
            if(sum==k)
            maxi=i+1;
            if(mp.find(sum-k)!=mp.end())
            maxi=max(maxi,i-mp[sum-k]);
            if(mp.find(sum)==mp.end())
            mp[sum]=i;
        }
        return maxi;
    }
};

File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.java
// User function Template for Java

class Solution {
    public int longestSubarray(int[] arr, int k) {
        // code here
        Map<Integer,Integer> mp = new HashMap<>();
        int sum = 0;
        int ans = 0;
        for(int i=0;i<arr.length;i++){
            sum+=arr[i];
            if(sum==k)
                ans = i+1;
            if(mp.containsKey(sum-k))
                ans = Math.max(ans,i-mp.get(sum-k));
            if(!mp.containsKey(sum))
                mp.put(sum,i);
        }
        return ans;
    }
}
